### ISSUE TITLE
backwards compatibility hazard when adding optional arguments to labelled functions

### ISSUE THREAD

**Issue by v-gb** [ocaml/ocaml#13253]
The following program fails to typecheck:

```ocaml
let foo ?tail:_ ~loc:(_: float) () = ()
let gen f x = f ~loc:1. x
let _foo = gen foo
(*
7 | let _foo = gen foo
                   ^^^
Error: This expression has type ?tail:'a -> loc:float -> unit -> unit
       but an expression was expected of type loc:float -> 'b -> 'c
*)
```

Having the typer accept that program would be both useful (for evolving APIs without breaking callers) and expected (because it does work if the other parameters aren't labelled, so it's very surprising to see non-orthogonality between labelling an argument and adding an optional argument).

Is there a principled reason why this can't work, or is it possible but simply not done ?

## More info

The same program but without labelling `~loc`, which the compiler accepts.

```ocaml
let foo ?tail:_ (_ : float) () = ()
let gen f x = f 1. x
let _foo = gen foo (* typechecks, the typer turns foo into foo ?tail:None or equivalent *)
```

This is consistent with what's described in the manual:

> Normally the compiler generates a type error if you attempt to pass to a function a parameter whose type is different from the expected one. However, in the specific case where the expected type is a non-labeled function type, and the argument is a function expecting optional parameters, the compiler will attempt to transform the argument to have it match the expected type, by passing None for all optional parameters.

Adding optional arguments (to an unlabelled function) to an API can break backwards compatibility in corner cases, but it's quite close to "just works". While writing this, I actually realize I have no clue where the compiler is eliminating optional arguments and why or why not. Apparently function arguments and datastructures work but not control flow for some reason:

```ocaml
# let f ?a () = ();;
# [ ignore; f ];;
- : (unit -> unit) list = [<fun>; <fun>]
# [ f; ignore ];;
Error: This expression has type 'a -> unit
       but an expression was expected of type ?a:'b -> unit -> unit
# type r = { a : unit -> unit };;
# { a = f };;
# if true then ignore else f;;
Error: This expression has type ?a:'a -> unit -> unit
       but an expression was expected of type 'b -> unit
# (if true then ignore else f : unit -> unit);;
Error: This expression has type ?a:'a -> unit -> unit
       but an expression was expected of type 'b -> unit
# if true then ignore else (f : unit -> unit);;
- : unit -> unit = <fun>
```

**Comment by samsa1** [comment]
I just looked rapidly into it and I'll write here what I understood for the record.

The function that does this elimination is `type_argument` and the easiest way to change the behavior to the semantic presented above would be to change the `no_labels` function by replacing `((=) Nolabel)` by `(fun l -> not (is_optional l))`.

https://github.com/ocaml/ocaml/blob/4e27e80af8653a2a5f3ea0cd16aeca841807206c/typing/typecore.ml#L5804-L5810

However doing the change as I just presented seems like a bad idea in my opinion because this would significantly impact the readability of the error message in the following example.
```ocaml
let foo ?loc:_ () = ()
let gen f x = f ~loc:1. x
let _foo = gen foo
(*
7 | let _foo = gen foo
                   ^^^
Error: This expression has type ?loc:'a -> unit -> unit
       but an expression was expected of type loc:'a -> 'b -> 'c
*)
```
will instead have the following error message :
```ocaml
(*
7 | let _foo = gen foo
                   ^^^
Error: This expression has type unit -> unit
       but an expression was expected of type loc:'a -> 'b -> 'c
*)
```
that most user would not understand. As such work would be needed to improve error messages if we want such a change of semantic.

**Comment by garrigue** [comment]
There is a principled reason why this is not allowed.
Namely, you cannot tell from the type of `gen` whether it is always going to apply `foo` in a way that discards the argument `?tail`.
In particular, `gen` could have been defined as:
```ocaml
let gen f : 'a -> 'b = f ~loc:1.0;;
```
which has exactly the same type, but does not apply `f` to an unlabelled argument, and as such cannot discard optional ones.

**Comment by samsa1** [comment]
Don't we already have this problem without labels ?

Namely :
```ocaml
let gen1 (f : unit -> unit) = f ()
let gen2 (f : unit -> unit) = let _ = f in ()
```
Both have the same type but only one applies `f` to an unlabelled argument. However both of them discard the optional argument.

**Comment by v-gb** [comment]
Thanks for the response.

I don't really follow the reasoning. Does the paper about labels talk about these difficulties? Maybe I should go take a look it.

My understanding is that labelled/optional arguments commute/are discarded only when typing an application with sufficiently known types. And optional parameters can get discarded based on their expected type in some places. The rest of the time, like during unification of function types, types have to match exactly.
So when you have:

```ocaml
  let gen f = f ~loc:1.0
```
you know from the type of `gen` that its parameter always receives `~loc` first. It seems to me that if you pass in:

```ocaml
  let foo ?tail:_ ~loc:(_ : float) () = ()
  let _ = gen foo
```

things behave fine if the typer turns `gen foo` into `gen (fun ~loc -> foo ?tail ~loc)`, the same thing that the typer would do if ~loc wasn't labelled. Perhaps it's more unexpected, because instead of discard the optional parameter, you could reorder the parameters so the optional parameter is not discarded:

```ocaml
  let foo2 ~loc:(_ : float) ?tail:_ () = ()
  let z = gen foo2 ?tail:None (* typechecks ok *)
```

Although considering that currying is not really compatible with labelled/optional arguments, I'm not sure it's very reasonable to write this sort of code.

**Comment by garrigue** [comment]
@samsa1 No, the implementation respects the semantics: the discarding is only done if the function is applied (and as many times as it is applied):
```ocaml
# let f (_ : unit -> unit) = ();;
val f : (unit -> unit) -> unit = <fun>
# let f2 (g : unit -> unit) = g (); g();;
val f2 : (unit -> unit) -> unit = <fun>
# let g ?(x = prerr_endline "Hello"; 3) () = ();;
val g : ?x:int -> unit -> unit = <fun>
# g ();;
Hello
- : unit = ()
# f g ;;
- : unit = ()
# f2 g;;
Hello
Hello
- : unit = ()
```
To be more precise, the problem when applying the original `gen` to `foo` is that the compiler does not know how to properly preserve the semantics through eta-expansion. Namely, it is possible that `gen` it just returning the function after a partial application to `loc`, and later this function is applied simultaneously to `()` and `tail`. In that case, the untyped semantics says that `tail` should not be discarded, but rather the given value of `tail` used:
```ocaml
let foo' = gen f in foo' ~tail:[] ()
```
This said, in this precise case, one could argue that since the type system does not allow this program, this does not really matter.
If we consider the following `foo`, this does matter:
```ocaml
let foo ?tail:_ ~loc:(_: float) () ~tail:_ = ()
```
Accepting the first one would probably mean that we also accept this one, and the meaning of `foo'` is clearly different from the untyped semantics. Finding precisely in which cases the typing is sufficient to prevent this is not so easy, so the typechecker stays on the conservative side.

@v-gb There is supposed to be also an untyped semantics of labelled applications, that does not depend on the types assigned to expressions. Typing only allows to optimize it. The paper supposed to describe it is:
https://www.math.nagoya-u.ac.jp/~garrigue/papers/ppl2001.pdf
However, it is no up-to-date with the current semantics of ocaml, as some design changes occurred after that.
My most recent attempt at giving a semantics (actually, multiple ones) can be found in the following PR comment:
https://github.com/ocaml/ocaml/pull/10653#issuecomment-1815914118
